### ISSUE TITLE
unit tests fail with python < 3.11

### ISSUE THREAD

**Issue by markstur** [instructlab/instructlab#814]
tox -e unit passes on MacOS with python3.11, but with 3.9 or 3.10 the following error happens:

```
>           self.assertEqual(result.exit_code, 0)
E           AssertionError: 1 != 0

tests/test_lab_train.py:341: AssertionError
```

**Comment by booxter** [comment]
@tiran let's continue the discussion here since the PR was closed and I am not sure we want to proceed with the code as was posted.

You suggested that instead of moving imports around, we can adopt `mock_mlx` or something similar. I don't think this mock function solves the same problem though. See below.

When I run test cases on python 3.9, I see that 1) the very first test case that mocks out a function, .e.g `make_data`, indeed sees MagicMock() in lab.train function (I can validate it with pdb / print / write to file); but 2) any consequent test cases that mock the same function (`make_data`) see unpatched version of the function in the same lab.train code. This looks like the first test case unpatches a global state that is reused for the second test case. `pytest` docs suggest that some class level state is shared between test cases of the same class, which suggests to me that global state is also shared (same interpreter.)

I also suspect there was some fix in unittest.mock (?) or CPython interpreter that unshared some of it, which makes the problem show for py3.10 and lower only. But I couldn't find anything of interest in 3.11 release notes.

BTW I also think that `mock_mlnx` is invalid - it executes mock.patch.dict when the decorator is applied, not when the test case is actually executed. I believe the correct implementation would look like:

```
def mock_mlx(f):
    """mlx is not available on Linux"""
    mlx_modules = {
        name: MagicMock()
        for name in ["mlx", "mlx.core", "mlx.nn", "mlx.optimizers", "mlx.utils"]
    }

    def func(*args, **kwargs):
        with patch.dict(sys.modules, mlx_modules):
            return f(*args, **kwargs)

    return func
```

(Side note: I couldn't make much progress here, partially because I struggle with the "unit" tests using complete `click` library to run functions. It would - in my mind - be better if unit tests actually trigger "units" - functions under test.)

**Comment by tiran** [comment]
+1 for continuing the discussion here.

That's an great finding. You are a genius! Yes, the mock function is wrong. I made a mistake when I designed the mock helper for `mlx` package. It doesn't work as intended. On Python 3.11 and 3.12, it kinda works and tests are passing. I have a local fix with a better and simpler implementation that uses a pytest fixture instead.

The current implementation of the unit test suite is someward problematic. It mixes `unittest.TestCase` style testing with `pytest` style testing. The combination can have bad side effects and cannot utilize the full potential of pytest. I recommend to embrace pytest and drop any use of the unittest package except for `unittest.mock`.

- do not subclass from `unittest.TestCase`
- replace `with self.assertRaises()` with `with pytest.raises()`
- replace `self.assert*()` with pytest-style bare `assert`
